pkg/master/controller.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package master
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/endpoint"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/namespace"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service"
+	servicecontroller "github.com/GoogleCloudPlatform/kubernetes/pkg/registry/service/allocator/controller"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	"github.com/golang/glog"
+)
+
+// Controller is the controller manager for the core bootstrap Kubernetes controller
+// loops, which manage creating the "kubernetes" and "kubernetes-ro" services, the "default"
+// namespace, and provide the IP repair check on service PortalIPs
+type Controller struct {
+	NamespaceRegistry namespace.Registry
+	ServiceRegistry   service.Registry
+	ServiceIPRegistry service.IPRegistry
+	EndpointRegistry  endpoint.Registry
+	PortalNet         *net.IPNet
+
+	PortalIPInterval time.Duration
+	EndpointInterval time.Duration
+
+	PublicIP net.IP
+
+	ServiceIP         net.IP
+	ServicePort       int
+	PublicServicePort int
+
+	ReadOnlyServiceIP         net.IP
+	ReadOnlyServicePort       int
+	PublicReadOnlyServicePort int
+
+	runner *util.Runner
+}
+
+// Start begins the core controller loops that must exist for bootstrapping
+// a cluster.
+func (c *Controller) Start() {
+	if c.runner != nil {
+		return
+	}
+
+	repair := servicecontroller.NewRepair(c.PortalIPInterval, c.ServiceRegistry, c.PortalNet, c.ServiceIPRegistry)
+
+	// run all of the controllers once prior to returning from Start.
+	if err := repair.RunOnce(); err != nil {
+		glog.Errorf("Unable to perform initial IP allocation check: %v", err)
+	}
+	if err := c.UpdateKubernetesService(); err != nil {
+		glog.Errorf("Unable to perform initial Kubernetes service initialization: %v", err)
+	}
+	if err := c.UpdateKubernetesROService(); err != nil {
+		glog.Errorf("Unable to perform initial Kubernetes RO service initialization: %v", err)
+	}
+
+	c.runner = util.NewRunner(c.RunKubernetesService, c.RunKubernetesROService, repair.RunUntil)
+	c.runner.Start()
+}
+
+// RunKubernetesROService periodically updates the kubernetes service
```

Comment:
Yes, although the code that was refactored was to use registries for all cases (now that we've correctly turned our registries into wrappers for source).  I will double check services though - they may be the last one that needs a conversion.

----- Original Message -----
> > +	// run all of the controllers once prior to returning from Start.
> > +	if err := repair.RunOnce(); err != nil {
> > +		glog.Errorf("Unable to perform initial IP allocation check: %v", err)
> > +	}
> > +	if err := c.UpdateKubernetesService(); err != nil {
> > +		glog.Errorf("Unable to perform initial Kubernetes service
> > initialization: %v", err)
> > +	}
> > +	if err := c.UpdateKubernetesROService(); err != nil {
> > +		glog.Errorf("Unable to perform initial Kubernetes RO service
> > initialization: %v", err)
> > +	}
> > +
> > +	c.runner = util.NewRunner(c.RunKubernetesService,
> > c.RunKubernetesROService, repair.RunUntil)
> > +	c.runner.Start()
> > +}
> > +
> > +// RunKubernetesROService periodically updates the kubernetes service
> 
> A bunch of the comments here and below got out of sync with the function
> names.
> 
> Is it safe to assume that the code was essentially unchanged from
> pkg/master/publish.go though?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7704/files#r29630786